runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java

```diff
@@ -31,17 +34,47 @@ public SequencerView(CorfuRuntime runtime) {
      * @return The first token retrieved.
      */
     public TokenResponse nextToken(Set<UUID> streamIDs, int numTokens) {
-        return layoutHelper(e -> CFUtils.getUninterruptibly(e.getPrimarySequencerClient()
-                .nextToken(streamIDs, numTokens)));
+        if (numTokens == 0) {
+            long tail = checkHints(streamIDs);
+            if (tail != Address.NOT_FOUND) {
+                // the request can be serviced from sequencer hints (omitting sequencer request)
+                long epoch = TransactionalContext.getCurrentContext().hintsEpoch;
```

Comment:
The sequencer view shouldn't be aware of the transactional context

Reply:
This is the best place to intercept the calls because its the least invasive. Let me know if you have other suggestions related to code org.

Reply:
TransactionalContext is a threadLocal . Does it matter software layering wise where it is accessed ? It seams like it should be cross cutting across the layers. 
If the code looks not clean what is the alternative that is being suggested ?

Reply:
Yes, why should the sequencer view (which should function only to get a view of the current sequencer) be aware or even be able to access the current transaction? It would not be reasonable to expect that the behavior of next token CHANGE depending on what transaction is currently running! And it should at least be documented, which doesn't happen here!

And as for implementation alternatives, one simple alternative would be to call ```nextToken()``` with all the streams, fix the sequencer logic to deal with such a request instead of implementing a new API, and to fix access inside of ```OptimisiticTransactionContext``` to use the hint by passing the hint instead of using snapshotTimestamp during the call to ```syncWithRetryUnsafe```. In stream view, change the behavior so that the sequencer is not called when maxGlobal is given and try resolving a backpointer first if not in the resolved queue.

But it seems that it would just make much more sense, since you're doing all this work, to just cache the entire table you transferred in the view itself, obviating the need for any of the mechanisms we just discussed.

Reply:
Its actually less code too. Moved the hints to OptimisiticTransactionContext and removed all the changes in SequencerView.